include/power/power.h

```diff
@@ -65,6 +65,29 @@ extern unsigned char pm_idle_exit_notify;
  * @{
  */
 
+/**
+ * Power management notifier struct
+ *
+ * This struct contains callbacks that are called when the target enters and
+ * exits power states.
```

Comment:
From what I can tell, this only affects transition from `PM_STATE_ACTIVE` to a different state (possibly excluding `PM_STATE_RUNTIME_IDLE`, it's not clear).  Thus all observable sequences are exactly:

0. start from `PM_STATE_ACTIVE`
1. enter state X
2. leave state X
3. back in `PM_STATE_ACTIVE`

The solution does *not* support:

0. start from `PM_STATE_ACTIVE`
1. enter state X
2. enter state Y
3. leave state Y
4. enter state Z
5. leave state Z
6. leave state X
7. back in `PM_STATE_ACTIVE`

where X might be `PM_STATE_RUNTIME_IDLE`, for example. Is that correct?

If so that's a pretty fundamental behavior that needs to be explained somewhere, either here or in the definition of the power states.  Documenting it here for now would be enough.

I say that because device power management will almost certainly not be that flat.

Reply:
@pabigot The solution does not change **when** the notifications are sent, I have not changed this logic yet. This is just a new way to send these notifications and I have already explained that. I'm splitting these changes in several pr because I can get early feedback and do the changes. As for example, your comments about concurrent access to this list and etc ... 

That is way better than expend a long time change everything, including board specific code and have to re-do it many times due fundamental mistakes in the base code like this.

Reply:
I *know* it's not changing anything.  I just want what it *does* to be documented, because it wasn't when this was originally added, so anybody using it has to dig to figure it out.

I'll accept this with the following change.

```
diff --git a/include/power/power.h b/include/power/power.h
index 11a64ebb94..69bfd993e7 100644
--- a/include/power/power.h
+++ b/include/power/power.h
@@ -71,6 +71,11 @@ extern unsigned char pm_idle_exit_notify;
  * This struct contains callbacks that are called when the target enters and
  * exits power states.
  *
+ * As currently implemented the entry callback is invoked when
+ * transitioning from PM_STATE_ACTIVE to another state, and the exit
+ * callback is invoked when transitioning from a non-active state to
+ * PM_STATE_ACTIVE.  This behavior may change in the future.
+ *
  * @note These callbacks can be called from the ISR of the event
  *       that caused the kernel exit from idling.
  */
```

Reply:
That is fair enough